### PR TITLE
Simplify HashMap Bucket interface

### PR DESCRIPTION
**Pull request by arthurprs**
> Simplify HashMap Bucket interface
> 
> * Store capacity_mask instead of capacity
> * Move bucket index into RawBucket
> * Valid bucket index is now always within [0..table_capacity)
> * Simplify iterators by moving logic into RawBuckets
> * Clone RawTable using RawBucket
> * Make retain aware of the number of elements

The idea was to put idx in RawBucket instead of the other Bucket types and simplify next() and prev() as much as possible. The rest was a side-effect of that change, except maybe the last 2.

This change makes iteration and other next/prev() heavy operations noticeably faster. Clone is way faster.

```
➜  hashmap2 git:(adapt) ✗ cargo benchcmp pre:: adp:: bench.txt 
 name                        pre:: ns/iter  adp:: ns/iter  diff ns/iter   diff % 
 clone_10_000                74,364         39,736              -34,628  -46.57% 
 grow_100_000                8,343,553      8,233,785          -109,768   -1.32% 
 grow_10_000                 817,825        723,958             -93,867  -11.48% 
 grow_big_value_100_000      18,418,979     17,906,186         -512,793   -2.78% 
 grow_big_value_10_000       1,219,242      1,103,334          -115,908   -9.51% 
 insert_1000                 74,546         58,343              -16,203  -21.74% 
 insert_100_000              6,743,770      6,238,017          -505,753   -7.50% 
 insert_10_000               798,079        719,123             -78,956   -9.89% 
 insert_1_000_000            275,215,605    266,975,875      -8,239,730   -2.99% 
 insert_int_bigvalue_10_000  1,517,387      1,419,838           -97,549   -6.43% 
 insert_str_10_000           316,179        278,896             -37,283  -11.79% 
 insert_string_10_000        770,927        747,449             -23,478   -3.05% 
 iter_keys_100_000           386,099        333,104             -52,995  -13.73% 
 iterate_100_000             387,320        355,707             -31,613   -8.16% 
 lookup_100_000              206,757        193,063             -13,694   -6.62% 
 lookup_100_000_unif         219,366        193,180             -26,186  -11.94% 
 lookup_1_000_000            206,456        205,716                -740   -0.36% 
 lookup_1_000_000_unif       659,934        629,659             -30,275   -4.59% 
 lru_sim                     20,194,334     18,442,149       -1,752,185   -8.68% 
 merge_shuffle               1,168,044      1,063,055          -104,989   -8.99%
```

Note 2: I may have messed up porting the diff, let's see what CI says.

### CODE REVIEW

File: src/libstd/collections/hash/table.rs [hunk 1]
```diff
@@ -126,9 +126,10 @@ unsafe impl<K: Send, V: Send> Send for RawTable<K, V> {}
 unsafe impl<K: Sync, V: Sync> Sync for RawTable<K, V> {}
 
 pub struct RawBucket<K, V> {
-    hash: *mut HashUint,
+    _hash: *mut HashUint,
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: Why was this renamed? If it was to disambiguate with the method, I think `hash_` would be more appropriate than `_hash`, since the standard practice is to use prefix `_` to denote "unused."

File: src/libstd/collections/hash/table.rs [hunk 2]
```diff
@@ -801,25 +784,31 @@ impl<K, V> RawTable<K, V> {
...
             size: 0,
             hashes: TaggedHashUintPtr::new(hashes),
             marker: marker::PhantomData,
         }
     }
 
-    fn first_bucket_raw(&self) -> RawBucket<K, V> {
-        let hashes_size = self.capacity * size_of::<HashUint>();
-        let pairs_size = self.capacity * size_of::<(K, V)>();
+    fn raw_bucket(&self) -> RawBucket<K, V> {
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: This seems like a worse name, since it isn't clear which raw bucket is returned. Maybe `first_raw_bucket` or `raw_bucket_first` or stick with the original `first_bucket_raw` (presumably paired with `bucket_at_raw`).

File: src/libstd/collections/hash/table.rs [hunk 3]
```diff
@@ -233,12 +231,14 @@ fn can_alias_safehash_as_hash() {
 }
 
 impl<K, V> RawBucket<K, V> {
-    unsafe fn offset(self, count: isize) -> RawBucket<K, V> {
-        RawBucket {
-            hash: self.hash.offset(count),
-            pair: self.pair.offset(count),
-            _marker: marker::PhantomData,
-        }
+    unsafe fn hash(&self) -> *mut HashUint {
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: Could you add a comment noting that these functions are unsafe because you can safely create out-of-bounds buckets?

File: src/libstd/collections/hash/table.rs [hunk 4]
```diff
@@ -126,9 +126,10 @@ unsafe impl<K: Send, V: Send> Send for RawTable<K, V> {}
 unsafe impl<K: Sync, V: Sync> Sync for RawTable<K, V> {}
 
 pub struct RawBucket<K, V> {
-    hash: *mut HashUint,
+    _hash: *mut HashUint,
     // We use *const to ensure covariance with respect to K and V
-    pair: *const (K, V),
+    _pair: *const (K, V),
+    idx: usize,
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: If this is always within `[0..table_capacity)`, could you add a comment mentioning that.

File: src/libstd/collections/hash/table.rs [hunk 5]
```diff
@@ -899,10 +886,8 @@ impl<K, V> RawTable<K, V> {
     /// Returns an iterator that copies out each entry. Used while the table
     /// is being dropped.
     unsafe fn rev_move_buckets(&mut self) -> RevMoveBuckets<K, V> {
-        let raw_bucket = self.first_bucket_raw();
         RevMoveBuckets {
-            raw: raw_bucket.offset(self.capacity as isize),
-            hashes_end: raw_bucket.hash,
+            raw: self.raw_bucket_at(self.capacity()),
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: You said in your description that bucket indices are in `[0..table_capacity)`, but this is one past that. Did you mean `[0..table_capacity]`?

File: src/libstd/collections/hash/table.rs [hunk 6]
```diff
@@ -938,29 +923,39 @@ impl<'a, K, V> Clone for RawBuckets<'a, K, V> {
...
+            elems_left: self.elems_left,
             marker: marker::PhantomData,
         }
     }
 }
 
 
 impl<'a, K, V> Iterator for RawBuckets<'a, K, V> {
-    type Item = RawBucket<K, V>;
+    type Item = (*mut HashUint, *mut (K, V));
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: It seems odd to have an iterator named `RawBuckets` that doesn't return `RawBucket`s.

File: src/libstd/collections/hash/table.rs [hunk 7]
```diff
@@ -938,29 +923,39 @@ impl<'a, K, V> Clone for RawBuckets<'a, K, V> {
...
 
-    fn next(&mut self) -> Option<RawBucket<K, V>> {
-        while self.raw.hash != self.hashes_end {
+    fn next(&mut self) -> Option<(*mut HashUint, *mut (K, V))> {
+        if self.elems_left == 0 {
+            return None;
+        }
+        loop {
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: Nit: could there be a blank line before this loop?

File: src/libstd/collections/hash/table.rs [hunk 8]
```diff
@@ -938,29 +923,39 @@ impl<'a, K, V> Clone for RawBuckets<'a, K, V> {
...
             }
         }
+    }
 
-        None
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        (self.elems_left, Some(self.elems_left))
+    }
+}
+impl<'a, K, V> ExactSizeIterator for RawBuckets<'a, K, V> {
```
Comment (gereeter) [inline src/libstd/collections/hash/table.rs]: Nit: could there be a blank line before this `impl`.

File: src/libstd/collections/hash/table.rs (hunk 5 above)
Comment (arthurprs) [inline src/libstd/collections/hash/table.rs]: This is intentionally set to one past the end for the reverse style iterator. I will add a comment line on that.

File: src/libstd/collections/hash/table.rs (hunk 6 above)
Comment (arthurprs) [inline src/libstd/collections/hash/table.rs]: Good point, hopefully that won't affect performance negatively.

File: src/libstd/collections/hash/table.rs (hunk 1 above)
Comment (arthurprs) [inline src/libstd/collections/hash/table.rs]: It's for disambiguation. I think we can come up with a better name instead, moving the _ to the end hurts the eye. Like `hash_start`, f`hash_base, ' first_hash`?

File: src/libstd/collections/hash/table.rs (hunk 2 above)
Comment (arthurprs) [inline src/libstd/collections/hash/table.rs]: I'll switch callers to `raw_bucket_at(0)`.

**Comment by arthurprs** [comment]
Thanks for the review @gereeter. I'll updated the PR later.

**Comment by arthurprs** [comment]
Rebased.

**Comment by alexcrichton** [comment]
cc @bluss, @pczarn

**Comment by arthurprs** [comment]
gentle ping

File: src/libstd/collections/hash/map.rs [hunk 9]
```diff
@@ -503,7 +498,7 @@ fn robin_hood<'a, K: 'a, V: 'a>(bucket: FullBucketMut<'a, K, V>,
         loop {
             displacement += 1;
             let probe = bucket.next();
-            debug_assert!(probe.index() != idx_end);
+            debug_assert!(probe.index() != start_index);
```
Comment (pczarn) [inline src/libstd/collections/hash/map.rs]: Is this invariant not as strict as the old one? Either way it's ok. You could add a comment that this assertion is simplified.

**Review by pczarn — APPROVED** [review]
Excellent. ~~Please make a rebase~~

File: src/libstd/collections/hash/map.rs (hunk 9 above)
Comment (arthurprs) [inline src/libstd/collections/hash/map.rs]: It's not. I'll updat the code to do the same check as before now.

**Comment by arthurprs** [comment]
Updated with @pczarn review.

**Comment by alexcrichton** [comment]
@bors: retry

* network failure https://github.com/rust-lang/rust/issues/40474
